Sources/Datadog/Tracing/Utils/MethodSwizzler.swift

```diff
@@ -6,56 +6,126 @@
 
 import Foundation
 
-internal enum SwizzlingError: LocalizedError {
-    case unknown
-    case classIsNotNSObjectSubclass(className: String)
-    case methodNotFound(selector: String, className: String)
-    case methodWasNotSwizzled(selector: String, className: String)
-}
-
 internal class MethodSwizzler {
```

Comment:
Looking at the `MethodSwizzler` interface, we can see that it's suffering from the lack of OOP:
```swift
class MethodSwizzler {
   func findMethodRecursively(with selector: Selector, in klass: AnyClass) -> FoundMethod?
   func currentImplementation<TypedIMP>(of found: FoundMethod) -> TypedIMP
   func originalImplementation<TypedIMP>(of found: FoundMethod) -> TypedIMP
   func swizzleIfNonSwizzled(foundMethod: FoundMethod, with implementation: @autoclosure () -> IMP) -> Bool
   func set(newIMP: IMP, for found: FoundMethod)
}
```
The `FoundMethod` appears in **all** internal method signatures. Using OOP, we can basically turn the `FoundMethod` into an object and simplify this code a lot. I gave it a try on [this branch](https://github.com/DataDog/dd-sdk-ios/blob/ncreated/RUMM-452-propose-swizzling-refactoring/Sources/Datadog/Tracing/Utils/MethodSwizzler.swift), resulting with half the code removed, no global `implementationCache` to maintain and no threading consideration.

The trick is to turn all above methods into instance methods on `FoundMethod` (I renamed it to `MethodSwizzler` as IMO it sounds more correct for the problem we solve):
```swift
class MethodSwizzler {
    /// The `Method` controlled by this swizzler.
    let method: Method

    static func `for`(selector: Selector, inClass class: AnyClass) -> MethodSwizzler?

    func originalImplementation<TypedIMP>() -> TypedIMP
    func currentImplementation<TypedIMP>() -> TypedIMP
    func swizzle(to newIMP: IMP)
    func swizzleIfNotSwizzled(to newIMP: IMP)
}
```

---

I think this also solves one problem that we will fall into in a future: If we need to swizzle the same method (e.g. `URLSession.dataTaskWithURLRequest`) in two places, e.g. in `TracingInterceptor` and for `RUMInterceptor`, the `implementationCache` hashing function will be colliding, because the `className & methodName` hash will be the same for swizzlings done in `TracingInterceptor` and `RUMInterceptor`.

My proposed way solves this, as the swizzling is identified by `MethodSwizzler` instance.

---

Last, but not least, because we will be swizzling from multiple places in the app (from different features), IMO the most easy and performant way of ensuring thread safety would be to share the same `swizzlingQueue` between all swizzlers (including `URLSessionSwizzler`). This way,
each of them can perform its logic sequentially on the same thread. This should be way more performant than wrapping each `swizzle()` in a critical section (1 critical section per Swizzler vs many critical sections per Swizzler). That's why I removed the `sync {}` method.

WDYT @buranmert ?

Reply:
#### `FoundMethod`

you can think of this type as an `IMP` with its `class`, as Cocoa doesn't provide such a type we introduced our own. i don't see any problem with using it in `internal` interface.? 🤔 

#### `impCache` and thread safety

this is incorrect.
you call that OOP but this type modifies the runtime under the hood. the runtime is a shared resource, we always should consider access from multiple threads.
so, we cannot bypass threading issues with the proposed approach.
and we lose the track of swizzlings we made at runtime without `impCache`, that would lead to swizzle the same class (eg: `__NSURLSessionDataTask`) over and over again ❌ 

the problem with the proposed approach is proxying a singleton (ie: runtime) with a non-singleton (ie: `MethodSwizzler`) ❌ 
from my own experience, that usually ends up bad.

#### `FoundMethod` collision

this is incorrect. `impCache` keeps original implementations of methods. if we swizzle the same method in the same class, that doesn't change its value in `impCache` and we still can get its original imp from the cache.
in short, we can swizzle the same methods multiple times ✅ 
plus, we can attempt to undo it if anything ever goes wrong during swizzling 🦺 

**_not in the scope of this PR_** but i think we should minimize the number of swizzling we make. if we know, at runtime `Tracing` and `RUM` both require swizzling IMO the best option would be 
1. swizzling once with multiple `requestInterceptors` or 
2. a special `Tracing&RUM` interceptor

#### Queue vs lock

i first considered a serial private queue as [`Firebase` did](https://github.com/firebase/firebase-ios-sdk/blob/master/GoogleUtilities/MethodSwizzler/GULSwizzler.m#L33)

but that didn't really make sense to me: 
we can do swizzling in a private queue, but that doesn't mean the swizzled method will be called in that private queue. what problem does a private queue solve in this case?
that's why i used `objc_sync_enter/exit` only to restrict access from multiple threads

i'm not sure if there is huge perf diff between those 2, i should benchmark them 🛠️

Reply:
> you call that OOP but this type modifies the runtime under the hood.

The principle of OOP is to use abstraction to hide the implementation details. In case of abstracting the swizzling logic, runtime modification is an implementation detail.

The [InterposeKit](https://github.com/steipete/InterposeKit) is a good example of abstracting the Swizzling by objects:
```swift
let swizzling = try Interpose(URLSession.self) {
    try $0.hook(#selector(URLSession.dataTaskWithURL) { 
        /* provide swizzled implementation */
    }
}

swizzling.revert() // unswizzle
```
I makes the swizzling conceptually easy to digest: _"this object manages one method swizzling"_. Similarly to what was proposed, the underlying swizzlings are modeled as objects: [`Task`](https://github.com/steipete/InterposeKit/blob/f8daba1194b3118b819261374e318c8a958975fa/Sources/InterposeKit/InterposeKit.swift#L95-L106), which validates it as a working solution to the problem.

But I don't insist on going into `InterposeKit's` design (I don't even don't know if it supports what we need). I only say that encapsulation with objects is more elegant and reusable, in contrast to direct manipulation on `IMPs`:
```swift
private static func newIMP(using interceptor: @escaping RequestInterceptor) throws -> IMP {
let newImp_URL = DataTask_URL_Completion.newIMP(
    using: interceptor,
    methodToSwizzle: method_dataTask_URL,
    methodToRedirect: method_dataTask_request
)
swizzler.set(newIMP: newImp_URL, for: method_dataTask_URL)

let newImp_request = DataTask_Request_Completion.newIMP(
    using: interceptor,
    methodToSwizzle: method_dataTask_request
)
swizzler.set(newIMP: newImp_request, for: method_dataTask_request)
```

But the way it was updated in recent commit is good enough for readability :+1:. Sadly, we won't be able to reuse it for RUM.

---

> `impCache` keeps original implementations of methods. if we swizzle the same method in the same class, that doesn't change its value in impCache and we still can get its original imp from the cache.

OK, so this is the part that I didn't get, but this is an implementation detail in any abstraction we take.

> if we know, at runtime Tracing and RUM both require swizzling IMO the best option would be
> 1. swizzling once with multiple requestInterceptors or
> 2. a special Tracing&RUM interceptor

The first option sounds fine. The second one will mix up the responsibility of shared interceptor, making it creak when testing (meaning a bad abstraction).

> but that didn't really make sense to me:
> we can do swizzling in a private queue, but that doesn't mean the swizzled method will be called in that private queue. what problem does a private queue solve in this case?
that's why i used objc_sync_enter/exit only to restrict access from multiple threads

Sure, we don't want to change the execution thread of swizzled methods.

The problem is not only **how to synchronize**, but more importantly, **where to synchronize**.

#### Where to synchronize?

There are two choices:
* in `URLSessionSwizzler`,
* in `MethodSwizzler` - the current solution.

If we synchronize in `URLSessionSwizzler`, the critical section will be executed `N * number of tasks` times and won't grow when the number of `swizzler` access grows:
```
URLSessionSwizzler.installSwizzling() { // <- one critical section access
    swizzleDataTaskWithURLRequest()
    swizzleDataTaskWithURL()
}

// later, each time when a `URLSessionTask` is created:

swizzleDataTaskResume() // <- one critical section access
```

If we synchronize in `MethodSwizzler`, it will be `3 * N * number of tasks` times and will grow by `N` with every new `swizzler` access we need:
```swift
URLSessionSwizzler.installSwizzling() {
    swizzleDataTaskWithURLRequest() {
        // ...
        swizzler.currentImplementation() // <- 1st critical section access
        // ...
        swizzler.set(newIMP: newImp, for: method) // <- 2nd critical section access
        // ...
    }
    swizzleDataTaskWithURL() {
        // ...
        swizzler.currentImplementation() // <- 3rd critical section access
        // ...
        swizzler.set(newIMP: newImp, for: method) // <- 4th critical section access
        // ...
    }
}

// later, each time when a `URLSessionTask` is created:

swizzleDataTaskResume() {
    // ...
    swizzler.findMethodRecursively() // <- 1st critical section access
    // ...
    swizzler.swizzleIfNonSwizzled // <- 2nd critical section access
    // ...
    swizzler.currentImplementation(of: foundMethod) // <- 3rd critical section access
    // ...
}
```

Although I skipped it, the impact of initial swizzlings is not negligible, as we may have many of them done in the RUM product.

#### How to synchronize?

There are two choices:
* sync,
* async.

If we do `sync`, we will block the application thread (almost always the main thread) for the time of critical section access. Given that we currently do 3 accesses per task, this might result in a visible FPS drop.

If we do `async`, we won't be dropping any application frames, as swizzling will be non-blocking. The only problem with `async` which I see is the `task.resume()` which might be called before we swizzle. That can be easily overcome by checking the `task.status` before swizzling.

#### Picking the synchronization tool

Only given "where" and "how" we can discuss the implementation of critical section. 

Current approach with `objc_sync_enter(self)` is **blocking**, so the main thread will be impacted. Yes, it is more performant than `DispatchQueue` according to [some benchmarks](https://www.cocoawithlove.com/blog/2016/06/02/threads-and-mutexes.html), but it doesn't give us the off-main-thread execution.

That's why IMO using a `DispatchQueue` is more relevant to our problem: conceptually easier, quite performant and makes async execution possible.

Reply:
in fact, we can reduce the number of critical region access with block-based approach 👍 
```swift
func swizzle<TypedIMP>(method: FoundMethod, impProvider: (currentImp: TypedIMP) -> IMP) {
  sync {
    ...
  }
}
```
that way `MethodSwizzler` provides `currentImplementation` within `impProvider` block
i couldn't do that before because we needed to get `currentImp` separately in order to do redirection in `dataTaskWithURL`
i'll make that change now ✅ 

Firebase is swizzling in its own private queue in `sync`, even they don't seem to cause perf issues (benchmarking should help anyway though)
`async` cannot be an option, the app needs to wait while the runtime is being changed 🤷‍♂️ 
`objc_sync_enter/exit` still looks like the best option to me ✅ 

If we model a single swizzling as an object:
1. what does it give us?
2. this is incorrect abstraction
let's say we made `swizzle1` and then `swizzle2`; if we now `swizzle1.revert` that will also revert `swizzle2`

p.s: this thread is too long to follow now, can we start a new one if we still have questions?

Reply:
> If we model a single swizzling as an object:
> 1. what does it give us?

First and crucial, it allows us to get rid of threading synchronization for `IMP` storage at all (!), because the `originalIMP` and `currentIMP` will be bound to the `MethodSwizzler` object, instead of being stored in a **global, shared** `implementationCache: [FoundMethod: IMP]` which requires synchronization (as "global + shared resources" usually do).

Second, it is much cleaner, as it enables the _"Is this method already swizzled?"_ reasoning on the right level of abstraction: inside the `URLSessionSwizzler`. This vs _"I will swizzle this method, but maybe it won't require swizzling and something internally will handle it"_.

Third, this is more performant, not only because it doesn't force the thread synchronization for **all swizzlings**, but also because it gives us the freedom to selectively decide which swizzlings we want to synchronize. Considering two swizzlers:
* `UIViewControllerSwizzler`
* `URLSessionSwizzler`

**Why would we need to synchronize the `UIViewController.viewWillLoad` swizzlings?** We can ask the users to call `Datadog.autoinstrumentViewControllers()` early on the main thread and this will be benefitial for both sides: zero thread-switch penalty in the running app and much easier threading model in the SDK code (well: single threading model).

**Why do we need to synchronize `dataTaskWithURL` and `dataTaskWithURLRequest` swizzlings in `URLSessionSwizzler`?** Again, asking the user to call `Datadog.autoinstrumentURLSession()` early on the main thread is a win-win. Things are different if it comes to `task.resume()` swizzling, as that one is applied later in the running app, so that one requires synchronization (and even this can be solved - what I describe later).

So, using object-oriented approach:
* is not forcing any thread-switch for **every swizzling** (no FPS drop),
* allows for selectively synchronizing **some swizzlings** if necessary (and so far we only have one).

---

Also, here's my proposal to do `task.resume()` swizzling on the main thread. Using the object-oriented approach, it might look this:
```swift
class URLSessionSwizzler {
    /// Object applying `URLSession.dataTaskWithURL()` swizzling.
    private let dataTaskWithURLSwizzling: DataTaskWithURLSwizzling
    
    /// Object applying `URLSession.dataTaskWithURLRequest()` swizzling.
    private let dataTaskWithRequestSwizzling: DataTaskWithURLRequestSwizzling

    /// Object applying `URLSessionTask.resume()` swizzling.
    private let dataTaskResumeSwizzling: DataTaskResumeSwizzling

    init() {
        do {
            // check if we can swizzle:
            self.dataTaskWithURLSwizzling = try DataTaskWithURLSwizzling()
            self.dataTaskWithRequestSwizzling = try DataTaskWithURLRequestSwizzling()
            self.dataTaskResumeSwizzling = try DataTaskResumeSwizzling()

            // swizzle:
            self.dataTaskWithURLSwizzling.apply()
            self.dataTaskWithRequestSwizzling.apply()
            self.dataTaskResumeSwizzling.apply()
        } catch {
            print("Sorry, we cannot swizzle `URLSession` because: \(error)")
        }
    }

    /// Applies the `URLSession.dataTaskWithURL()` swizzling.
    private struct DataTaskWithURLSwizzling {
        private let swizzling: MethodSwizzling

        init() throws {
            try MethodSwizzling(
                class: URLSession.self,
                selector: #selector(URLSession.dataTask(with:completionHandler:) as ...
            )
        }

        func apply() {
            swizzling.addImplementation { ... in
                // ...
            }
        }
    }

    /// Applies the `URLSession.dataTaskWithURLRequest()` swizzling.
    private struct DataTaskWithURLRequestSwizzling {
        private let swizzling: MethodSwizzling

        init() throws {
            try MethodSwizzling(
                class: URLSession.self,
                selector: #selector(URLSession.dataTask(with:completionHandler:) as ...
            )
        }

        func apply() {
            swizzling.addImplementation { ... in
                // ...
            }
        }
    }

    /// Applies the `URLSessionTask.resume()` swizzling.
    private struct DataTaskResumeSwizzling {
        private let swizzling: MethodSwizzling

        init() throws {
            let dummyURLSession = URLSession()
            let task = dummyURLSession.dataTaskWithRequests(...)

            self.swizzling = try MethodSwizzling(
                class: object_getClass(task),
                selector: #selector(URLSessionTask.resume as ...
            )

            task.cancel()
        }

        func apply() {
            swizzling.addImplementation { ... in
                // ...
            }
        }
    }
}
```
```swift
class UIViewControllerSwizzling {
    /// Object applying `UIViewController.viewWillAppear()` swizzling.
    private let viewWillAppearSwizzling: ViewWillAppearSwizzling
    
    /// Object applying `UIViewController.viewDidDisappear()` swizzling.
    private let viewDidDisappearSwizzling: ViewDidDisappearSwizzling

    init() {
        // ...
    }

    /// Applies the `UIViewController.viewWillAppear()` swizzling.
    private struct ViewWillAppearSwizzling {
        // ...
    }    

    /// Applies the `UIViewController.viewDidDisappear()` swizzling.
    private struct ViewDidDisappearSwizzling {
        // ...
    }    
}
```

> 2. this is incorrect abstraction let's say we made swizzle1 and then swizzle2; if we now swizzle1.revert that will also revert swizzle2

First, we don't need to unswizzle our swizzlings in production code. And punishing the customer apps for unswizzling need in unit tests is not a way to go.

Second, all this can be done in object-oriented way. I don't see why `MethodSwizzler` object couldn't track all swizzlings that were done to a given selector? The same way as we track them in `implementationCache: [FoundMethod: IMP]`. The difference is crucial: it will track it locally and only when necessary, instead of using global, shared and synchronized cache. This abstraction is not making it impossible in any way. Again, [`InterposeKit`](https://github.com/steipete/InterposeKit) uses something similar, with success.

---

Wrapping up:
* object-oriented abstraction lets us pick the execution thread per swizzling,
* global-cache abstraction forces the same execution thread for all swizzlings.

And less key factor: the OO abstraction I proposed is half the number of code less than we need for global-cache.

Reply:
summary: this discussion went quite long because we were not talking about the same thing.

we have 2 main use cases: 
1. static: swizzling `URLSession` only once
2. dynamic: swizzling every `URLSessionTask` created (if needed)
(i made up the names "static" and "dynamic", i hope they make sense)

this PR covers both of them, static and dynamic. 
@ncreated argues that dynamic swizzling can be covered per-case basis as a special case.

i assume this is resolved by now